### PR TITLE
Use pkg-config to find Flann

### PR DESCRIPTION
**Pull request by jspricke**

### CODE REVIEW

**Comment by jspricke** [comment]
Hey guys, this is an change I recently introduced in Debian, some details are covered here:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=911022#14
Would be great to have it in the 1.9 release (#2399) to make sure PCL compiles on future Debian and Ubuntu distros. If someone could test this on other systems, that would be great.

P.S.: I just merged this from the simpler patch I did for Debian and the old version, will hopefully have time tomorrow to actually test it ;). Just wanted to hand it in so you are aware of it already. Feel free to fix my typos ;).

**Comment by claudiofantacci** [comment]
Thanks for the PR, I can try to test this under Windows with vcpkg in the next days 👍

File: cmake/Modules/FindFLANN.cmake [hunk 1]
```diff
@@ -19,41 +19,41 @@ endif(FLANN_USE_STATIC)
...
 find_package(PkgConfig QUIET)
 if (FLANN_FIND_VERSION)
-    pkg_check_modules(PC_FLANN flann>=${FLANN_FIND_VERSION})
+    pkg_check_modules(FLANN flann>=${FLANN_FIND_VERSION})
 else(FLANN_FIND_VERSION)
-    pkg_check_modules(PC_FLANN flann)
+    pkg_check_modules(FLANN flann)
 endif(FLANN_FIND_VERSION)
 
-set(FLANN_DEFINITIONS ${PC_FLANN_CFLAGS_OTHER})
```
Comment (SergioRAgostinho) [inline cmake/Modules/FindFLANN.cmake]: Is FLANN no longer exporting definitions?

Comment (taketwo) [reply cmake/Modules/FindFLANN.cmake]: Ubuntu 16.04, `$ pkg-config --cflags-only-other flann` yields empty string.

**Comment by SergioRAgostinho** [comment]
Confirmed the same on OS X.

**Comment by taketwo** [comment]
I'm taking back my words, it's not fine, more details later.

**Comment by taketwo** [comment]
In my case `pkgconfig` succeeds in find the library, so commands in the `if(NOT FLANN_FOUND)` branch are never executed. This means that `FLANN_INCLUDE_DIRS` variable is not set. Because of that, PCL thinks that FLANN is not missing and does not build any of the modules that depend on FLANN (including transitive dependencies). You can observe this in the Travis build, which succeeded in 10 minutes since it did not need to build much.

In the case when `pkgconfig` succeeds a variable called `FLANN_INCLUDEDIR` is set (at least on Ubuntu 16.04). So we need to assign `FLANN_INCLUDE_DIRS` to it.

**Comment by SergioRAgostinho** [comment]
This behavior definitely needs to be patched. The
find_package_handle_standard_args needs to move outside the conditional
branch.

**Comment by jspricke** [comment]
@taketwo I added a workaround the the `FLANN_FOUND`. The `FLANN_INCLUDE_DIRS` should be set by `pkg_check_modules` (It's probably empty on Ubuntu). Can you try?
@SergioRAgostinho `find_package_handle_standard_args` is not needed.

**Comment by SergioRAgostinho** [comment]
My comment was based on a practice which goes as following: when something doesn't fail and it should have, you first write code to ensure it fails and then you prevent the failure from happening.

Furthermore this patch 12dc5a2abb79335b78d8c6fb756d88b715ef6aff will appear slightly cryptic to someone who just sees this file. `FLANN_FOUND` is already being implicitly evaluated as `TRUE` in the previous line. I tried to look for other occurrences of `FLANN_FOUND` in the repo but saw no explicit comparison of the type
```
FLANN_FOUND STREQUAL TRUE
```

**Comment by taketwo** [comment]
There was no problem with `FLANN_FOUND` variable. The problem is that `FLANN_INCLUDE_DIRS` is **not** set by `pkg_check_modules` on Ubuntu 16.04. Instead, a variable called `FLANN_INCLUDEDIR` is set.

**Comment by jspricke** [comment]
I tested it on Ubuntu and it works.
Read here if you want to know more about `FLANN_INCLUDE_DIRS` https://cmake.org/cmake/help/v3.5/module/FindPkgConfig.html (hint, it's all fine).

**Comment by taketwo** [comment]
Is this still WIP, do you plan to change anything else?

File: cmake/Modules/FindFLANN.cmake [hunk 2]
```diff
@@ -19,43 +19,44 @@ endif(FLANN_USE_STATIC)
...
 
-include(FindPackageHandleStandardArgs)
-find_package_handle_standard_args(FLANN DEFAULT_MSG FLANN_LIBRARY FLANN_INCLUDE_DIR)
-
-mark_as_advanced(FLANN_LIBRARY FLANN_LIBRARY_DEBUG FLANN_INCLUDE_DIR)
+    mark_as_advanced(FLANN_LIBRARY FLANN_LIBRARY_DEBUG FLANN_INCLUDE_DIR)
+endif()
 
 if(FLANN_FOUND)
+  set(FLANN_FOUND TRUE)
```
Comment (UnaNancyOwen) [inline cmake/Modules/FindFLANN.cmake]: Why set <code>TRUE</code> to <code>FLANN_FOUND</code>? It already looks like <code>TRUE</code> in here.

Comment (jspricke) [reply cmake/Modules/FindFLANN.cmake]: It's set to 1 when you use `pkg_check_modules` and later we do `STREQUAL TRUE`

**Comment by jspricke** [comment]
@taketwo it's fine with me (if it works for everyone). I would love to get rid of the `FLANN_FOUND TRUE` thing, but I fear that would change too much of our cmake internals this late in the release process.
Thanks everyone for helping with getting this into the release.

File: cmake/Modules/FindFLANN.cmake (hunk 2 above)
Comment (UnaNancyOwen) [reply cmake/Modules/FindFLANN.cmake]: Oh, I get it. Thanks! 👍

Comment (jspricke) [reply cmake/Modules/FindFLANN.cmake]: Sounds like this really needs some documentation, did so :).

**Comment by claudiofantacci** [comment]
It works under Windows 10 with `vcpkg`:
```
FLANN found (include: C:/Users/cfantacci/vcpkg/installed/x64-windows/include, lib: optimized;C:/Users/cfantacci/vcpkg/installed/x64-windows/lib/flann_cpp.lib;debug;C:/Users/cfantacci/vcpkg/installed/x64-windows/debug/lib/flann_cpp-gd.lib)
```

**Comment by taketwo** [comment]
@jspricke Actually, do you see any reason for this test you referenced:

https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/cmake/pcl_targets.cmake#L113

Why isn't the first part of the condition sufficient? Looking at the Git history, this was added as a single commit and does not provide any motivation. Maybe we just drop it? Or better after the release?

**Comment by claudiofantacci** [comment]
@taketwo, my two cents here: there are part of CMake codes where some variables are set/evaluated `ON`/`OFF` (Bool), other where they are considered as literals `TRUE`/`FALSE` (string). My guess is that this may depends on the package configurations file. It may be risky to change this.

**Comment by SergioRAgostinho** [comment]
I'm pro getting rid of that second condition after the release. One more
task for the CMake rework.

**Comment by jspricke** [comment]
I looked into this a little more and I think `OR` should be `AND`:
```
-            if(NOT ${EXT_DEP_FOUND} OR (NOT (${EXT_DEP_FOUND} STREQUAL "TRUE")))
+            if(NOT ${EXT_DEP_FOUND} AND (NOT (${EXT_DEP_FOUND} STREQUAL "TRUE")))
```

**Comment by jspricke** [comment]
I updated the logic, please squash before merging.

File: cmake/Modules/FindFLANN.cmake [hunk 3]
```diff
@@ -19,43 +19,44 @@ endif(FLANN_USE_STATIC)
...
 
-include(FindPackageHandleStandardArgs)
-find_package_handle_standard_args(FLANN DEFAULT_MSG FLANN_LIBRARY FLANN_INCLUDE_DIR)
-
-mark_as_advanced(FLANN_LIBRARY FLANN_LIBRARY_DEBUG FLANN_INCLUDE_DIR)
+    mark_as_advanced(FLANN_LIBRARY FLANN_LIBRARY_DEBUG FLANN_INCLUDE_DIR)
+endif()
 
 if(FLANN_FOUND)
+  # pkg_check_modules sets FLANN_FOUND to 1, set it to TRUE here, as we do STREQUAL TRUE later
```
Comment (taketwo) [inline cmake/Modules/FindFLANN.cmake]: You forgot to remove this comment.

**Comment by jspricke** [comment]
Updated, thx.

**Comment by SergioRAgostinho** [comment]
My output on OS X also seems to be fine. 👍
```
-- Checking for module 'flann>=1.7.0'
--   Found flann, version 1.9.1
-- FLANN found (include: /usr/local/Cellar/flann/1.9.1_3/include, lib: flann;flann_cpp)
```

**Pull request merged**